File: scripts/skill_stats.py
#!/usr/bin/env python3
from __future__ import annotations

import json
from collections import defaultdict
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
STORE = ROOT / 'data' / 'adaptive-skills-store.json'


def main() -> None:
    data = json.loads(STORE.read_text(encoding='utf-8'))
    grouped: dict[str, dict[str, int]] = defaultdict(lambda: {'success': 0, 'failure': 0})

    for record in data.get('records', []):
        skill = str(record.get('skill', 'unknown'))
        if record.get('success') is True:
            grouped[skill]['success'] += 1
        else:
            grouped[skill]['failure'] += 1

    rows = []
    for skill, stats in grouped.items():
        attempts = stats['success'] + stats['failure']
        success_rate = 0.0 if attempts == 0 else stats['success'] / attempts
        rows.append({
            'skill': skill,
            'attempts': attempts,
            'success': stats['success'],
            'failure': stats['failure'],
            'success_rate': round(success_rate, 3),
        })

    rows.sort(key=lambda row: (-row['attempts'], row['skill']))
    print(json.dumps(rows, indent=2))


if __name__ == '__main__':
    main()

File: scripts/skill-stats.mjs
#!/usr/bin/env node

import fs from "node:fs";
import path from "node:path";

const root = path.resolve(import.meta.dirname, "..");
const storePath = path.join(root, "data", "adaptive-skills-store.json");
const store = JSON.parse(fs.readFileSync(storePath, "utf8"));
const grouped = new Map();

for (const record of store.records || []) {
  const skill = String(record.skill || "unknown");
  if (!grouped.has(skill)) {
    grouped.set(skill, { success: 0, failure: 0 });
  }
  const stats = grouped.get(skill);
  if (record.success === true) {
    stats.success += 1;
  } else {
    stats.failure += 1;
  }
}

const rows = [...grouped.entries()].map(([skill, stats]) => {
  const attempts = stats.success + stats.failure;
  return {
    skill,
    attempts,
    success: stats.success,
    failure: stats.failure,
    success_rate: attempts === 0 ? 0 : Number((stats.success / attempts).toFixed(3)),
  };
});

rows.sort((a, b) => b.attempts - a.attempts || a.skill.localeCompare(b.skill));
console.log(JSON.stringify(rows, null, 2));
